pkg/kademlia/dht.go

```diff
@@ -5,37 +5,178 @@ package kademlia
+
 // Kademlia is an implementation of kademlia adhering to the DHT interface.
 type Kademlia struct {
+	rt             RoutingTable
+	bootstrapNodes []proto.Node
+	ip             string
+	port           string
+	stun           bool
+	dht            *bkad.DHT
+}
+
+// NewKademlia returns a newly configured Kademlia instance
+func NewKademlia(bootstrapNodes []proto.Node, ip string, port string, stun bool) Kademlia {
+	bb := convertProtoNodes(bootstrapNodes)
+	bdht, _ := bkad.NewDHT(&bkad.MemoryStore{}, &bkad.Options{
+		ID:             []byte(newID()),
+		IP:             ip,
+		Port:           port,
+		BootstrapNodes: bb,
+	})
+
+	rt := RouteTable{
+		ht:  bdht.HT,
+		dht: bdht,
+	}
+
+	return Kademlia{
+		rt:             rt,
+		bootstrapNodes: bootstrapNodes,
+		ip:             ip,
+		port:           port,
+		stun:           stun,
+		dht:            bdht,
+	}
 }
 
-// GetNodes returns all nodes from a starting node up to a maximum limit
-func (k Kademlia) GetNodes(ctx context.Context, start string, limit int) ([]overlay.Node, error) {
-	return []overlay.Node{}, nil
+// GetNodes returns all nodes from a starting node up to a maximum limit stored in the local routing table
+func (k Kademlia) GetNodes(ctx context.Context, start string, limit int) ([]proto.Node, error) {
+	nn, err := k.dht.FindNodes(ctx, start, limit)
+	if err != nil {
+		return []proto.Node{}, err
+	}
+	return convertNetworkNodes(nn), nil
 }
 
 // GetRoutingTable provides the routing table for the Kademlia DHT
 func (k Kademlia) GetRoutingTable(ctx context.Context) (RoutingTable, error) {
-	return RouteTable{}, nil
+	return RouteTable{
+		ht:  k.dht.HT,
+		dht: k.dht,
+	}, nil
 }
 
 // Bootstrap contacts one of a set of pre defined trusted nodes on the network and
 // begins populating the local Kademlia node
 func (k Kademlia) Bootstrap(ctx context.Context) error {
-	return nil
+	if err := k.dht.CreateSocket(); err != nil {
+		return err
+	}
+
+	go k.dht.Listen()
+
+	return k.dht.Bootstrap()
+
 }
 
 // Ping checks that the provided node is still accessible on the network
-func (k Kademlia) Ping(ctx context.Context, node overlay.Node) (overlay.Node, error) {
-	return overlay.Node{}, nil
+func (k Kademlia) Ping(ctx context.Context, node proto.Node) (proto.Node, error) {
+	n, ok := convert(node).(*bkad.NetworkNode)
+	if !ok {
+		return proto.Node{}, NodeErr.New("unable to convert to expected type")
+	}
+	ok, err := k.dht.Ping(n)
+	if err != nil {
+		return proto.Node{}, err
+	}
+	if !ok {
+		return proto.Node{}, NodeErr.New("node unavailable")
+	}
+	return node, nil
 }
 
 // FindNode looks up the provided NodeID first in the local Node, and if it is not found
 // begins searching the network for the NodeID. Returns and error if node was not found
-func (k Kademlia) FindNode(ctx context.Context, ID NodeID) (overlay.Node, error) {
-	return overlay.Node{}, nil
+func (k Kademlia) FindNode(ctx context.Context, ID NodeID) (proto.Node, error) {
+	nodes, err := k.dht.FindNode([]byte(ID))
```

Comment:
weird that you can't just tell it exact match. we should fix the lower level lib eventually

Reply:
we're doing extra work to make up for the extra work it's doing

Reply:
Agreed. After demos, I want to write our own Kademlia implementation. This is purely to get us something up and running so we can get all the pieces talking

Reply:
:+1: